### PR TITLE
Metadata handling updates

### PR DESCRIPTION
**Pull request by jasongrout**
This PR implements the changes from #52.

TODO:
- [x] Make metadata use local member when instantiating a new HasTraits
- [x] deprecation warning
- [x] add tag method
- [x] change tests
- [x] fix configure code
- [x] Don't do the deprecation warning if just specifying the `help` argument

### CODE REVIEW

File: traitlets/traitlets.py [hunk 1]
```diff
@@ -64,7 +64,8 @@
 #-----------------------------------------------------------------------------
 # Basic classes
 #-----------------------------------------------------------------------------
-
+import warnings
+warnings.filterwarnings('always')
```
Comment (jasongrout) [inline traitlets/traitlets.py]: Remove these in the final PR

File: traitlets/traitlets.py [hunk 2]
```diff
@@ -356,13 +357,14 @@ def __init__(self, default_value=Undefined, allow_none=None, **metadata):
             )
 
         if len(metadata) > 0:
+            warn("metadata should be set by ...", DeprecationWarning, stacklevel=2)
```
Comment (jasongrout) [inline traitlets/traitlets.py]: Fix this message with whatever is decided

Comment (ellisonbg) [inline traitlets/traitlets.py]: Let's reach consensus on the discussion in #52 before going too much
further on the implementation.

On Fri, Jul 17, 2015 at 12:48 PM, Jason Grout <notifications@github.com>
wrote:

> In traitlets/traitlets.py
> <https://github.com/ipython/traitlets/pull/53#discussion_r34924703>:
>
> > @@ -356,13 +357,14 @@ def __init__(self, default_value=Undefined, allow_none=None, **metadata):
> >              )
> >
> >          if len(metadata) > 0:
> > +            warn("metadata should be set by ...", DeprecationWarning, stacklevel=2)
>
> Fix this message with whatever is decided
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ipython/traitlets/pull/53/files#r34924703>.
>



-- 
Brian E. Granger
Cal Poly State University, San Luis Obispo
@ellisonbg on Twitter and GitHub
bgranger@calpoly.edu and ellisonbg@gmail.com

Comment (jasongrout) [inline traitlets/traitlets.py]: Yep, we're on the same page.  That's why I put up the note above.

**Comment by jasongrout** [comment]
I think this is ready for review.

**Comment by ellisonbg** [comment]
Great, thanks!

On Wed, Jul 22, 2015 at 12:25 PM, Jason Grout <notifications@github.com>
wrote:

> I think this is ready for review.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ipython/traitlets/pull/53#issuecomment-123834439>.
>



-- 
Brian E. Granger
Cal Poly State University, San Luis Obispo
@ellisonbg on Twitter and GitHub
bgranger@calpoly.edu and ellisonbg@gmail.com

**Comment by Carreau** [comment]
You emacs seem to make the test fail.

**Comment by jasongrout** [comment]
hehe...couldn't be emacs, must be human error :)

**Comment by jasongrout** [comment]
All the above PRs to the various repos are works-in-progress.

**Comment by jasongrout** [comment]
Yep, you're right.  I was over-eager in my replacements early on, including here.

**Comment by jasongrout** [comment]
Actually, we should probably make help a first-class argument of the constructor on this PR too.  So this is still WIP.

**Comment by jasongrout** [comment]
rebased on master...

File: traitlets/traitlets.py [hunk 3]
```diff
@@ -359,13 +358,21 @@ def __init__(self, default_value=Undefined, allow_none=None, read_only=None, **m
             )
 
         if len(metadata) > 0:
+            warn("metadata should be set using the .tag() method, e.g., Int().tag(key1='value1', key2='value2')", DeprecationWarning, stacklevel=2)
```
Comment (minrk) [inline traitlets/traitlets.py]: This should include the metadata keys being set. Most people who encounter this message won't have any idea what values are metadata and what aren't.

Comment (jasongrout) [inline traitlets/traitlets.py]: Good call!

**Comment by jasongrout** [comment]
We'll do the help promotion in another PR.  It's sufficiently invasive and different that it deserves its own discussion.

**Comment by jasongrout** [comment]
Okay, this is ready again for review.

File: traitlets/config/tests/test_configurable.py [hunk 4]
```diff
@@ -368,7 +368,7 @@ class SomeSingleton(SingletonConfigurable):
             pass
 
         class DefaultConfigurable(Configurable):
-            a = Integer(config=True)
+            a = Integer().tag(config=True)
             def _config_default(self):
                 if SomeSingleton.initialized():
                     return SomeSingleton.instance().config
```
Comment (minrk) [inline traitlets/config/tests/test_configurable.py]: Can we keep some tests that pass `config=True` in the constructor to ensure we are preserving the deprecated behavior? I know it's tested at the lower level in traitlets, but I'd like to keep one or two here, as well.

Comment (jasongrout) [inline traitlets/config/tests/test_configurable.py]: Sure. Which test(s) do you want?  I'll copy it and check for the deprecation warning too.

Comment (minrk) [inline traitlets/config/tests/test_configurable.py]: Just one or two of the simplest existing tests. All I really care about is keeping track of `a = Trait(config=True)` still working and loading its value from the config object.

Comment (jasongrout) [inline traitlets/config/tests/test_configurable.py]: Okay, then https://github.com/jasongrout/traitlets/commit/eab41d651c828a7f62131adb589f91c32d2336b5 should be sufficient (i.e., the commit below).

**Comment by jdfreder** [comment]
Will this be released in the next traitlet minor?  Trying to decide when to merge Jason's widget PR - and for what version of widgets.

**Comment by SylvainCorlay** [comment]
I guess that we can require the next minor traitlets release in ipywidgets as soon as it is out.

**Pull request merged**